File: 0742-closest-leaf-in-a-binary-tree/0742-closest-leaf-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findClosestLeaf(self, root: Optional[TreeNode], k: int) -> int:
        self.start = None
        graph = collections.defaultdict(list)

        def dfs(node, parent, val):
            if node:
                if node.val == val:
                    self.start = node
                if parent:
                    graph[node].append(parent)
                    graph[parent].append(node)
                dfs(node.left, node, val)
                dfs(node.right, node, val)

        dfs(root, None, k)
        
        q = collections.deque([self.start])
        seen = set(q)
        
        while q:
            node = q.popleft()
            if node:
                if node.left is None and node.right is None:
                    return node.val
                for nei in graph[node]:
                    if nei not in seen:
                        seen.add(nei)
                        q.append(nei)
        
        return 0

File: 0742-closest-leaf-in-a-binary-tree/0742-closest-leaf-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*, vector<TreeNode*>> graph;
    TreeNode* start = nullptr;
    int findClosestLeaf(TreeNode* root, int k) {
        dfs(root, nullptr, k);
        queue<TreeNode*> q;
        q.push(start);
        unordered_set<TreeNode*> seen;
        while(!q.empty()) {
            int size = q.size();
            while(size--) {
                TreeNode* cur = q.front(); q.pop();
                seen.insert(cur);
                if(!cur->left && !cur->right) return cur->val;
                for(TreeNode* nei : graph[cur])
                    if (!seen.count(nei)) q.push(nei);
            }
        }
        return 0;
    }
    
    void dfs(TreeNode* node, TreeNode* parent, int val) {
        if(!node) return;
        if(node->val == val) start = node;
        if(parent) {
            graph[node].push_back(parent);
            graph[parent].push_back(node);
        }
        if(node->left) dfs(node->left, node, val);
        if(node->right) dfs(node->right, node, val);
    }
};